### ISSUE TITLE
2.0.1 - test fail with SQLCE provider

### ISSUE THREAD

**Issue by ErikEJ** [aspnet/EntityFrameworkCore#10177]
This change was added to 2.0.1:

https://github.com/aspnet/EntityFrameworkCore/commit/b59da98ee113500bb89ff5677f887a792eaa8d3f

I am seeing this test error

```
Test 'Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.Include_reference_collection_order_by_reference_navigation' failed:
	Assert.Equal() Failure
Expected: 3
Actual:   11
	at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQueryTestBase`2.ComplexNavigationsQueryResultAsserter.AssertLevel1(Level1 expected, Level1 actual, IEnumerable`1 expectedIncludes)
	at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQueryTestBase`2.ComplexNavigationsQueryResultAsserter.AssertElement[TElement](TElement expected, TElement actual, IEnumerable`1 expectedIncludes)
	at Microsoft.EntityFrameworkCore.TestUtilities.QueryResultAsserter.AssertCollection[TElement](IEnumerable`1 expected, IEnumerable`1 actual, IEnumerable`1 expectedIncludes)
	at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQueryTestBase`2.ComplexNavigationsQueryResultAsserter.AssertCollection[TElement](IEnumerable`1 expected, IEnumerable`1 actual, IEnumerable`1 expectedIncludes)
```

### Further technical details
EF Core version: 2.0.1-rtm-203
Database Provider: SQLCE
Operating system: Win 10
IDE: Vs 2017 15.3

**Comment by ajcvickers** [comment]
@ErikEJ What does the generated SQL for CE look like?

**Comment by ajcvickers** [comment]
@ErikEJ Also, link above doesn't seem to resolve--can you provide a good link to the change?

**Comment by ErikEJ** [comment]
Link fixed - SQL coming up!

**Comment by ErikEJ** [comment]
```


Test 'Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.Include_reference_collection_order_by_reference_navigation' failed:
	Assert.Equal() Failure
Expected: 3
Actual:   11
	at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQueryTestBase`2.ComplexNavigationsQueryResultAsserter.AssertLevel1(Level1 expected, Level1 actual, IEnumerable`1 expectedIncludes)
	at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQueryTestBase`2.ComplexNavigationsQueryResultAsserter.AssertElement[TElement](TElement expected, TElement actual, IEnumerable`1 expectedIncludes)
	at Microsoft.EntityFrameworkCore.TestUtilities.QueryResultAsserter.AssertCollection[TElement](IEnumerable`1 expected, IEnumerable`1 actual, IEnumerable`1 expectedIncludes)
	at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQueryTestBase`2.ComplexNavigationsQueryResultAsserter.AssertCollection[TElement](IEnumerable`1 expected, IEnumerable`1 actual, IEnumerable`1 expectedIncludes)

Output from Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.Include_reference_collection_order_by_reference_navigation:
  Entity Framework Core 2.0.1-rtm-203 initialized 'ComplexNavigationsContext' using provider 'EntityFrameworkCore.SqlServerCompact40' with options: SensitiveDataLoggingEnabled MaxBatchSize=1
  
  Compiling query model: 
  'from Level1 l1 in 
      (from Level1 l1 in DbSet<Level1>
      select [l1]).Include("OneToOne_Optional_FK.OneToMany_Optional")
  order by [l1].OneToOne_Optional_FK.Id asc
  select [l1]'
  
  Including navigation: '[l1].OneToOne_Optional_FK.OneToMany_Optional'
  
  Optimized query model: 
  'from Level1 l1 in DbSet<Level1>
  join Level2 l1.OneToOne_Optional_FK in DbSet<Level2>
  on (Nullable<int>)Property([l1], "Id") equals Property([l1.OneToOne_Optional_FK], "Level1_Optional_Id") into l1.OneToOne_Optional_FK_group
  from Level2 l1.OneToOne_Optional_FK in 
      (from Level2 l1.OneToOne_Optional_FK_groupItem in [l1.OneToOne_Optional_FK_group]
      select [l1.OneToOne_Optional_FK_groupItem]).DefaultIfEmpty()
  order by (int)[l1.OneToOne_Optional_FK]?.Id asc
  select Level1 _Include(
      queryContext: queryContext, 
      entity: [l1], 
      included: new object[]{ [l1.OneToOne_Optional_FK] }, 
      fixup: (QueryContext queryContext | Level1 entity | Object[] included) => 
      {
          return !(bool ReferenceEquals(included[0], null)) ? 
          {
              entity.OneToOne_Optional_FK = (Level2)included[0]
              ((Level2)included[0]).OneToOne_Optional_FK_Inverse = entity
              return Void queryContext.QueryBuffer.IncludeCollection(
                  includeId: 0, 
                  navigation: Level2.OneToMany_Optional, 
                  inverseNavigation: Level3.OneToMany_Optional_Inverse, 
                  targetEntityType: EntityType: Level3, 
                  clrCollectionAccessor: ClrICollectionAccessor<Level2, ICollection<Level3>, Level3>, 
                  inverseClrPropertySetter: ClrPropertySetter<Level3, Level2>, 
                  tracking: False, 
                  instance: (Level2)included[0], 
                  valuesFactory: () => 
                      from Level3 l1.OneToOne_Optional_FK.OneToMany_Optional in DbSet<Level3>
                      join AnonymousObject _l1.OneToOne_Optional_FK in 
                          (from Level1 l1 in DbSet<Level1>
                          join Level2 l1.OneToOne_Optional_FK in DbSet<Level2>
                          on (Nullable<int>)Property([l1], "Id") equals Property([l1.OneToOne_Optional_FK], "Level1_Optional_Id") into l1.OneToOne_Optional_FK_group
                          from Level2 l1.OneToOne_Optional_FK in 
                              (from Level2 l1.OneToOne_Optional_FK_groupItem in [l1.OneToOne_Optional_FK_group]
                              select [l1.OneToOne_Optional_FK_groupItem]).DefaultIfEmpty()
                          select new AnonymousObject(new object[]{ (object)EF.Property(?[l1.OneToOne_Optional_FK]?, "Id") })).Distinct()
                      on Property([l1.OneToOne_Optional_FK.OneToMany_Optional], "OneToMany_Optional_InverseId") equals (Nullable<int>)object [_l1.OneToOne_Optional_FK].GetValue(0)
                      order by object [_l1.OneToOne_Optional_FK].GetValue(0) asc
                      select [l1.OneToOne_Optional_FK.OneToMany_Optional])
          }
           : default(Void)
      }
  )'
  
  (QueryContext queryContext) => IEnumerable<Level1> _InterceptExceptions(
      source: IEnumerable<Level1> _Select(
          source: IEnumerable<TransparentIdentifier<Level1, Level2>> _ShapedQuery(
              queryContext: queryContext, 
              shaperCommandContext: SelectExpression: 
                  SELECT [l1].[Id], [l1].[Date], [l1].[Name], [l1].[OneToMany_Optional_Self_InverseId], [l1].[OneToMany_Required_Self_InverseId], [l1].[OneToOne_Optional_SelfId], [l1.OneToOne_Optional_FK].[Id], [l1.OneToOne_Optional_FK].[Date], [l1.OneToOne_Optional_FK].[Level1_Optional_Id], [l1.OneToOne_Optional_FK].[Level1_Required_Id], [l1.OneToOne_Optional_FK].[Name], [l1.OneToOne_Optional_FK].[OneToMany_Optional_InverseId], [l1.OneToOne_Optional_FK].[OneToMany_Optional_Self_InverseId], [l1.OneToOne_Optional_FK].[OneToMany_Required_InverseId], [l1.OneToOne_Optional_FK].[OneToMany_Required_Self_InverseId], [l1.OneToOne_Optional_FK].[OneToOne_Optional_PK_InverseId], [l1.OneToOne_Optional_FK].[OneToOne_Optional_SelfId]
                  FROM [Level1] AS [l1]
                  LEFT JOIN [Level2] AS [l1.OneToOne_Optional_FK] ON [l1].[Id] = [l1.OneToOne_Optional_FK].[Level1_Optional_Id]
                  ORDER BY [l1.OneToOne_Optional_FK].[Id], 
              shaper: TypedCompositeShaper<BufferedOffsetEntityShaper<Level1>, Level1, BufferedOffsetEntityShaper<Level2>, Level2, TransparentIdentifier<Level1, Level2>>), 
          selector: (TransparentIdentifier<Level1, Level2> t1) => Level1 _Include(
              queryContext: queryContext, 
              entity: t1.Outer, 
              included: new object[]{ t1.Inner }, 
              fixup: (QueryContext queryContext | Level1 entity | Object[] included) => 
              {
                  return !(bool ReferenceEquals(included[0], null)) ? 
                  {
                      entity.OneToOne_Optional_FK = (Level2)included[0]
                      ((Level2)included[0]).OneToOne_Optional_FK_Inverse = entity
                      return Void queryContext.QueryBuffer.IncludeCollection(
                          includeId: 0, 
                          navigation: Level2.OneToMany_Optional, 
                          inverseNavigation: Level3.OneToMany_Optional_Inverse, 
                          targetEntityType: EntityType: Level3, 
                          clrCollectionAccessor: ClrICollectionAccessor<Level2, ICollection<Level3>, Level3>, 
                          inverseClrPropertySetter: ClrPropertySetter<Level3, Level2>, 
                          tracking: False, 
                          instance: (Level2)included[0], 
                          valuesFactory: () => IEnumerable<Level3> _ShapedQuery(
                              queryContext: queryContext, 
                              shaperCommandContext: SelectExpression: 
                                  SELECT [l1.OneToOne_Optional_FK.OneToMany_Optional].[Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Level2_Optional_Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Level2_Required_Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Name], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_Self_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Required_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Required_Self_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToOne_Optional_PK_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToOne_Optional_SelfId]
                                  FROM [Level3] AS [l1.OneToOne_Optional_FK.OneToMany_Optional]
                                  INNER JOIN (
                                      SELECT DISTINCT [l1.OneToOne_Optional_FK0].[Id]
                                      FROM [Level1] AS [l10]
                                      LEFT JOIN [Level2] AS [l1.OneToOne_Optional_FK0] ON [l10].[Id] = [l1.OneToOne_Optional_FK0].[Level1_Optional_Id]
                                  ) AS [t] ON [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_InverseId] = [t].[Id]
                                  ORDER BY [t].[Id], 
                              shaper: BufferedEntityShaper<Level3>))
                  }
                   : default(Void)
              }
          )), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: Unhandled parameter: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l1].[Id], [l1].[Date], [l1].[Name], [l1].[OneToMany_Optional_Self_InverseId], [l1].[OneToMany_Required_Self_InverseId], [l1].[OneToOne_Optional_SelfId], [l1.OneToOne_Optional_FK].[Id], [l1.OneToOne_Optional_FK].[Date], [l1.OneToOne_Optional_FK].[Level1_Optional_Id], [l1.OneToOne_Optional_FK].[Level1_Required_Id], [l1.OneToOne_Optional_FK].[Name], [l1.OneToOne_Optional_FK].[OneToMany_Optional_InverseId], [l1.OneToOne_Optional_FK].[OneToMany_Optional_Self_InverseId], [l1.OneToOne_Optional_FK].[OneToMany_Required_InverseId], [l1.OneToOne_Optional_FK].[OneToMany_Required_Self_InverseId], [l1.OneToOne_Optional_FK].[OneToOne_Optional_PK_InverseId], [l1.OneToOne_Optional_FK].[OneToOne_Optional_SelfId]
  FROM [Level1] AS [l1]
  LEFT JOIN [Level2] AS [l1.OneToOne_Optional_FK] ON [l1].[Id] = [l1.OneToOne_Optional_FK].[Level1_Optional_Id]
  ORDER BY [l1.OneToOne_Optional_FK].[Id]
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l1.OneToOne_Optional_FK.OneToMany_Optional].[Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Level2_Optional_Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Level2_Required_Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Name], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_Self_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Required_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Required_Self_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToOne_Optional_PK_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToOne_Optional_SelfId]
  FROM [Level3] AS [l1.OneToOne_Optional_FK.OneToMany_Optional]
  INNER JOIN (
      SELECT DISTINCT [l1.OneToOne_Optional_FK0].[Id]
      FROM [Level1] AS [l10]
      LEFT JOIN [Level2] AS [l1.OneToOne_Optional_FK0] ON [l10].[Id] = [l1.OneToOne_Optional_FK0].[Level1_Optional_Id]
  ) AS [t] ON [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_InverseId] = [t].[Id]
  ORDER BY [t].[Id]
  
  A data reader was disposed.
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.

```

**Comment by smitpatel** [comment]
@ErikEJ - The SQL looks correct comparing with what SqlServer generates.

Can you run query separately in linq & with EF Core to compare results manually? (essentially take the test and re-write it without using AssertQuery pattern). That could help you debug why the ordering got messed up somehow.

When I looked at results in SqlServer, the ordering was by left joined table's Id (with nulls being first) and l1.Id was secondary ordering due to the fact that SqlServer data is PK sorted for tests.

**Comment by ErikEJ** [comment]
>>  linq & with EF Core

@smitpatel  Not sure what you mean by that?

**Comment by smitpatel** [comment]
EF Query
```C#
using (var context = CreateContext()) {
    var result = context.Level1s.Include(l1 => l1.OneToOne_Optional_FK.OneToMany_Optional).OrderBy(l1 => l1.OneToOne_Optional_FK.Id).ToList();
}
```
Linq query
```C#
using (var context = CreateContext()) {
    context.Level2s.ToList();
    context.Level3s.ToList();
    var query = context.Level1s.ToList()..OrderBy(l1 => l1.OneToOne_Optional_FK.Id).ToList();
```
If you write both of above in same test, then you can evaluate both the results during debugging and figure out what is going on.

The second query loads dbsets because of no Include on Linq

**Comment by ErikEJ** [comment]
Result of EF query:

```
SELECT [l1.OneToOne_Optional_FK.OneToMany_Optional].[Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Level2_Optional_Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Level2_Required_Id], [l1.OneToOne_Optional_FK.OneToMany_Optional].[Name], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_Self_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Required_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Required_Self_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToOne_Optional_PK_InverseId], [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToOne_Optional_SelfId]
  FROM [Level3] AS [l1.OneToOne_Optional_FK.OneToMany_Optional]
  INNER JOIN (
      SELECT DISTINCT [l1.OneToOne_Optional_FK0].[Id]
      FROM [Level1] AS [l10]
      LEFT JOIN [Level2] AS [l1.OneToOne_Optional_FK0] ON [l10].[Id] = [l1.OneToOne_Optional_FK0].[Level1_Optional_Id]
  ) AS [t] ON [l1.OneToOne_Optional_FK.OneToMany_Optional].[OneToMany_Optional_InverseId] = [t].[Id]
  ORDER BY [t].[Id]

```
Result rows (3 rows):

Id	Level2_Optional_Id	Level2_Required_Id	Name	OneToMany_Optional_InverseId	OneToMany_Optional_Self_InverseId	OneToMany_Required_InverseId	OneToMany_Required_Self_InverseId	OneToOne_Optional_PK_InverseId	OneToOne_Optional_SelfId
2	NULL	9	L3 02	1	NULL	1	1	NULL	9
6	NULL	5	L3 06	1	NULL	1	5	NULL	NULL
10	NULL	1	L3 10	1	NULL	1	9	NULL	NULL

**Comment by ErikEJ** [comment]
Output from second, which fails with null ref exception:
```

Test 'Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.TestIt' failed: System.NullReferenceException : Object reference not set to an instance of an object.
	ComplexNavigationsQuerySqlCeTest.cs(46,0): at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.<>c.<TestIt>b__4_0(Level1 l1)
	at System.Linq.EnumerableSorter`2.ComputeKeys(TElement[] elements, Int32 count)
	at System.Linq.EnumerableSorter`1.Sort(TElement[] elements, Int32 count)
	at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__1.MoveNext()
	at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
	at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
	ComplexNavigationsQuerySqlCeTest.cs(46,0): at Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.TestIt()

Output from Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.TestIt:
  Entity Framework Core 2.0.1-rtm-203 initialized 'ComplexNavigationsContext' using provider 'EntityFrameworkCore.SqlServerCompact40' with options: SensitiveDataLoggingEnabled MaxBatchSize=1
  
  Compiling query model: 
  'from Level2 <generated>_0 in DbSet<Level2>
  select [<generated>_0]'
  
  Optimized query model: 
  'from Level2 <generated>_0 in DbSet<Level2>
  select [<generated>_0]'
  
  (QueryContext queryContext) => IEnumerable<Level2> _InterceptExceptions(
      source: IEnumerable<Level2> _ShapedQuery(
          queryContext: queryContext, 
          shaperCommandContext: SelectExpression: 
              SELECT [l].[Id], [l].[Date], [l].[Level1_Optional_Id], [l].[Level1_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
              FROM [Level2] AS [l], 
          shaper: BufferedEntityShaper<Level2>), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l].[Id], [l].[Date], [l].[Level1_Optional_Id], [l].[Level1_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
  FROM [Level2] AS [l]
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Compiling query model: 
  'from Level3 <generated>_0 in DbSet<Level3>
  select [<generated>_0]'
  
  Optimized query model: 
  'from Level3 <generated>_0 in DbSet<Level3>
  select [<generated>_0]'
  
  (QueryContext queryContext) => IEnumerable<Level3> _InterceptExceptions(
      source: IEnumerable<Level3> _ShapedQuery(
          queryContext: queryContext, 
          shaperCommandContext: SelectExpression: 
              SELECT [l].[Id], [l].[Level2_Optional_Id], [l].[Level2_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
              FROM [Level3] AS [l], 
          shaper: BufferedEntityShaper<Level3>), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l].[Id], [l].[Level2_Optional_Id], [l].[Level2_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
  FROM [Level3] AS [l]
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Compiling query model: 
  'from Level1 <generated>_0 in DbSet<Level1>
  select [<generated>_0]'
  
  Optimized query model: 
  'from Level1 <generated>_0 in DbSet<Level1>
  select [<generated>_0]'
  
  (QueryContext queryContext) => IEnumerable<Level1> _InterceptExceptions(
      source: IEnumerable<Level1> _ShapedQuery(
          queryContext: queryContext, 
          shaperCommandContext: SelectExpression: 
              SELECT [l].[Id], [l].[Date], [l].[Name], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_SelfId]
              FROM [Level1] AS [l], 
          shaper: BufferedEntityShaper<Level1>), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l].[Id], [l].[Date], [l].[Name], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_SelfId]
  FROM [Level1] AS [l]
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.

0 passed, 1 failed, 0 skipped, took 2,69 seconds (xUnit.net 2.2.0 build 3545).


```

**Comment by smitpatel** [comment]
My bad in translating queries on the fly. 
Use this ordering `.OrderBy(l1 => MaybeScalar<int>(l1.OneToOne_Optional_FK, () => l1.OneToOne_Optional_FK.Id))` for linq query. The OrderBy is not null safe so NRE is expected.

**Comment by ErikEJ** [comment]
I cannot find "MaybeScalar" ??

**Comment by ErikEJ** [comment]
Result of LINQ to Objects test:

```
Output from Microsoft.EntityFrameworkCore.Query.ComplexNavigationsQuerySqlCeTest.TestIt:
  Entity Framework Core 2.0.1-rtm-203 initialized 'ComplexNavigationsContext' using provider 'EntityFrameworkCore.SqlServerCompact40' with options: SensitiveDataLoggingEnabled MaxBatchSize=1
  
  Compiling query model: 
  'from Level2 <generated>_0 in DbSet<Level2>
  select [<generated>_0]'
  
  Optimized query model: 
  'from Level2 <generated>_0 in DbSet<Level2>
  select [<generated>_0]'
  
  (QueryContext queryContext) => IEnumerable<Level2> _InterceptExceptions(
      source: IEnumerable<Level2> _ShapedQuery(
          queryContext: queryContext, 
          shaperCommandContext: SelectExpression: 
              SELECT [l].[Id], [l].[Date], [l].[Level1_Optional_Id], [l].[Level1_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
              FROM [Level2] AS [l], 
          shaper: BufferedEntityShaper<Level2>), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l].[Id], [l].[Date], [l].[Level1_Optional_Id], [l].[Level1_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
  FROM [Level2] AS [l]
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Compiling query model: 
  'from Level3 <generated>_0 in DbSet<Level3>
  select [<generated>_0]'
  
  Optimized query model: 
  'from Level3 <generated>_0 in DbSet<Level3>
  select [<generated>_0]'
  
  (QueryContext queryContext) => IEnumerable<Level3> _InterceptExceptions(
      source: IEnumerable<Level3> _ShapedQuery(
          queryContext: queryContext, 
          shaperCommandContext: SelectExpression: 
              SELECT [l].[Id], [l].[Level2_Optional_Id], [l].[Level2_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
              FROM [Level3] AS [l], 
          shaper: BufferedEntityShaper<Level3>), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l].[Id], [l].[Level2_Optional_Id], [l].[Level2_Required_Id], [l].[Name], [l].[OneToMany_Optional_InverseId], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_PK_InverseId], [l].[OneToOne_Optional_SelfId]
  FROM [Level3] AS [l]
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Compiling query model: 
  'from Level1 <generated>_0 in DbSet<Level1>
  select [<generated>_0]'
  
  Optimized query model: 
  'from Level1 <generated>_0 in DbSet<Level1>
  select [<generated>_0]'
  
  (QueryContext queryContext) => IEnumerable<Level1> _InterceptExceptions(
      source: IEnumerable<Level1> _ShapedQuery(
          queryContext: queryContext, 
          shaperCommandContext: SelectExpression: 
              SELECT [l].[Id], [l].[Date], [l].[Name], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_SelfId]
              FROM [Level1] AS [l], 
          shaper: BufferedEntityShaper<Level1>), 
      contextType: TestModels.ComplexNavigationsModel.ComplexNavigationsContext, 
      logger: DiagnosticsLogger<Query>, 
      queryContext: queryContext)
  
  Opening connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Opened connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='0']
  SELECT [l].[Id], [l].[Date], [l].[Name], [l].[OneToMany_Optional_Self_InverseId], [l].[OneToMany_Required_Self_InverseId], [l].[OneToOne_Optional_SelfId]
  FROM [Level1] AS [l]
  
  A data reader was disposed.
  
  Closing connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.
  
  Closed connection to database 'ComplexNavigations.sdf' on server 'ComplexNavigations.sdf'.

1 passed, 0 failed, 0 skipped, took 4,36 seconds (xUnit.net 2.2.0 build 3545).


```

**Comment by ErikEJ** [comment]
Anything else I can provide?

**Comment by smitpatel** [comment]
`MayBeScalar` implementation is available in source code of test base. 
Query plans are no good here. You need to debug the test and manually inspect results to figure out why ordering is different on SQL CE compared to LINQ. This is something provider specific issue.

**Comment by ErikEJ** [comment]
@smitpatel  This is fixed in 2.1 preview1 🎉
